### PR TITLE
[back] feat: add the suggestions/tocompare API

### PR DESCRIPTION
**Pull request by GresilleSiffle**
**related issues** #1785 

---

### Description

This PR adds the `/me/suggestions/{poll}/tocompare/` API.

This endpoint aims to suggest to the logged-in user a list of entities for comparison. In the future updates, the namespace `/me/suggestions/` could be extended to include new endpoints with different purposes:

```
# a list of entities to watch / to read / etc.
/me/suggestions/{poll}/towatch/

# a list of comparisons (or a single comparison) to make (to connect the personal graph, the collective graph, etc.)
/me/suggestions/{poll}/comparisons/
```

#### The strategies

The added endpoint is flexible and can suggest entities using different strategies. It's possible to add new strategy by creating a class inheriting from the abstract base class `ContributionSuggestionStrategy` and adding it to the API view. The default strategy that comes with this PR, called "classic", has several goals:
- be relatively simple
- return a list of entities the user "may want" to compare
- slightly mimic the behaviour of the front end's Auto button (while correcting its limitations)

The classic strategy is simple because it is based on simple heuristics: entities the user may want to compare are entities he/she started comparing or are in his/her rate-later list.

We suppose that the entities the user may want to compare in priority are those he/she started to compare few times, but not enough regarding the setting `rate_later__auto_remove`. We also suppose that the rate-later is a also good place to find those kind of entities, but we consider the rate-later slightly less reliable because depending on the users' habits it can be used as a watch-later list instead.

Finally, some flaws and limitations of the current frond end's Auto button have been fixed:

(a) We consider the compared pool as slightly more reliable than the rate-later list. The Auto button returns 75 % of entities from the rate-later list, the API returns 16 / 20 of entities from a combination of compared list and rate-later list.

(b) The complete compared list and rate-later list are queried for a better randomness, instead of just 100 enties.

(c) The recommended entities that represent 4 / 20 of the results come from the last month recommendations. Entities compared more than `rate_later__auto_remove` are excluded for a better experience. **Future improvement:** use the user's preferred languages.

**to-do**

*serializer*
- the new serializers looks very generic and may be redundant with others
  - [x] if relevant, move them to the correct file and/or rename them

*API endpoint*
- [x] choose a name (random/tocompare or suggestions/tocompare or something else)

*EntityToCompare view*
- [x] document the route
- [x] add the parameter strategy (default ToCompareStrategy.CLASSIC)
- [x] handle empty results returned by the strategy

*ClassicEntitySuggestionStrategy*
- [x] implement the pool: rate-later
- [x] implement the pool: reco last month
  - [x] use the setting `rate_later__auto_remove`
- [x] implement the pool: reco all time
  - [x] use the setting `rate_later__auto_remove`
- [x] ~add randomness to determine the selected pool~ all pool are selected to build a list of entities

*test*
- [x] add a test case for the strategy
- [x] add a test case for the API

**future improvements**

Make the `ClassicEntitySuggestionStrategy` behave differently for new and advanced users. The user profile can be determined by his/her number of comparisons, a low number indicates a new users, and high number indicates an advance users.

The pool rate-later could be replaced by "tutorial" entities for new users. A new additional pool "to connect" could be added for advanced users.

### Checklist

- [x] I added the related issue(s) id in the related issues section (if any)
  - if not, delete the related issues section
- [x] I described my changes and my decisions in the PR description
- [x] I read the development guidelines of the [CONTRIBUTING.md][development-guidelines]
- [x] The tests pass and have been updated if relevant
- [x] The code quality check pass

[development-guidelines]: https://github.com/tournesol-app/tournesol/blob/main/CONTRIBUTING.md#development-guidelines

### CODE REVIEW

File: backend/tournesol/lib/suggestions/strategies/tocompare/classic.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+        free_slots = self.max_suggestions - len(results)
+
+        if free_slots > 0:
+            last_resort = self._ids_from_pool_reco_all_time(results)
+            results += last_resort[:free_slots]
+
+        return Entity.objects.filter(id__in=results).with_prefetched_poll_ratings(
+            poll_name=poll.name
+        )
```
Comment (amatissart) [inline backend/tournesol/lib/suggestions/strategies/tocompare/classic.py:251]: Should we shuffle the list of entities when using `id__in`? The order of the results returned by the database is not deterministic, but probably not very random either.

**Review by amatissart — COMMENTED** [review]
I have minor comments about the implementation, but overall the approach looks good to me :+1:

File: backend/tournesol/views/suggestions/to_compare.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+from rest_framework import generics
+from rest_framework.permissions import IsAuthenticated
+
+from tournesol.lib.suggestions.strategies import ClassicEntitySuggestionStrategy
+from tournesol.serializers.suggestion import EntityToCompare
+from tournesol.views import PollScopedViewMixin
+
+
+class ToCompareStrategy(models.TextChoices):
+    CLASSIC = "Classic"
```
Comment (amatissart) [inline backend/tournesol/views/suggestions/to_compare.py:12]: I would find more usual to use lowercase in URL parameters
```suggestion
    CLASSIC = "classic"
```

File: backend/tournesol/lib/suggestions/strategies/tocompare/classic.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        user = self.user
+
+        max_threshold = user.settings.get(poll.name, {}).get(
+            "rate_later__auto_remove", RATE_LATER_AUTO_REMOVE_DEFAULT
+        )
+
+        return (
+            ContributorRating.objects.filter(poll=poll, user=user)
+            .select_related("entity")
+            .filter(**entity_filters)
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/tournesol/lib/suggestions/strategies/tocompare/classic.py]: Fixed by https://github.com/tournesol-app/tournesol/pull/1809/commits/c6ed71169f399ba29e9aa0002f747cd17633e4c9

File: backend/tournesol/lib/suggestions/strategies/tocompare/classic.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        return (
+            EntityPollRating.objects.filter(
+                poll=poll,
+                sum_trust_scores__gte=settings.RECOMMENDATIONS_MIN_TRUST_SCORES,
+                tournesol_score__gt=settings.RECOMMENDATIONS_MIN_TOURNESOL_SCORE,
+            )
+            .select_related("entity")
+            .filter(**entity_filters)
+            .exclude(entity_id__in=exclude_ids)
+            .values_list("entity_id", flat=True)
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/tournesol/lib/suggestions/strategies/tocompare/classic.py]: Fixed by https://github.com/tournesol-app/tournesol/pull/1809/commits/c6ed71169f399ba29e9aa0002f747cd17633e4c9

File: backend/tournesol/entities/video.py [hunk 5]
```diff
@@ -38,6 +38,10 @@ def get_allowed_meta_order_fields(cls) -> List[str]:
         """
         return ["duration", "publication_date"]
 
+    @classmethod
+    def get_filter_date_field(cls):
+        return "metadata__publication_date"
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/tournesol/entities/video.py:43]: Fixed by https://github.com/tournesol-app/tournesol/pull/1809/commits/3ad545c0e14c6818f287016bffce76bcb8e5c131

File: backend/tournesol/lib/suggestions/strategies/tocompare/classic.py (hunk 1 above)
Comment (GresilleSiffle) [reply backend/tournesol/lib/suggestions/strategies/tocompare/classic.py:244]: Exactly, it's better to shuffle the results.

Comment (GresilleSiffle) [reply backend/tournesol/lib/suggestions/strategies/tocompare/classic.py]: Fixed by https://github.com/tournesol-app/tournesol/pull/1809/commits/733e0ac5f0536c74fed0c8fc0dd979ad18e59f87

File: backend/tournesol/tests/test_api_suggestions_tocompare.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        self.poll2 = PollWithCriteriasFactory.create()
+
+        self.base_url = f"/users/me/suggestions/{self.poll1.name}/tocompare/"
+
+        self.user1 = UserFactory(username=self._username1)
+        self.user2 = UserFactory(username=self._username2)
+
+    def test_authentication_required(self):
+        """
+        Anonymous users cannot access their ratings' sub-samples.
```
Comment (GresilleSiffle) [inline backend/tournesol/tests/test_api_suggestions_tocompare.py:74]:
```suggestion
        Anonymous users cannot be suggested videos to compare.
```

File: backend/tournesol/tests/lib/suggestions/test_tocompare_classic.py [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+
+        # [WHEN] the filter `excluded_ids` is provided,
+        # [THEN] excluded entity ids should not be returned.
+        results = self.strategy._get_recommendations({}, exclude_ids=recent_entities)
+        self.assertEqual(len(results), 10)
+        self.assertTrue(set(results) == set(past_entities[:10]))
+
+    def test_get_compared_sufficiently(self):
+        """
+        The method `_ids_from_pool_compared` should return entity ids that
```
Comment (GresilleSiffle) [inline backend/tournesol/tests/lib/suggestions/test_tocompare_classic.py:126]:
```suggestion
        The method `_get_compared_sufficiently` should return entity ids that
```

**Review by amatissart — APPROVED** [review]

**Pull request merged**